php-checks/src/main/java/org/sonar/php/checks/MissingMethodVisibilityCheck.java

```diff
@@ -20,81 +20,79 @@
 @SqaleConstantRemediation("2min")
-public class MissingMethodVisibilityCheck extends SquidCheck<LexerlessGrammar> {
+public class MissingMethodVisibilityCheck extends PHPVisitorCheck {
 
-  private static final Set<PHPKeyword> VISIBILITIES = ImmutableSet.of(
-    PHPKeyword.PRIVATE,
-    PHPKeyword.PROTECTED,
-    PHPKeyword.PUBLIC);
+  public static final String KEY = "S1784";
+  private static final String MESSAGE = "Explicitly mention the visibility of this %s \"%s\".";
+
+  private static final Set<String> VISIBILITIES = ImmutableSet.of(
+    PHPKeyword.PRIVATE.getValue(),
+    PHPKeyword.PROTECTED.getValue(),
+    PHPKeyword.PUBLIC.getValue());
 
   @Override
-  public void init() {
-    subscribeTo(
-      PHPGrammar.METHOD_DECLARATION);
+  public void visitClassDeclaration(ClassDeclarationTree tree) {
+    super.visitClassDeclaration(tree);
+
+    if (tree.is(Kind.CLASS_DECLARATION)) {
+
+      for (ClassMemberTree member : tree.members()) {
+
+        if (member.is(Kind.METHOD_DECLARATION)) {
+          checkMethod((MethodDeclarationTree) member, tree.name().text());
+        }
+      }
+    }
   }
 
-  @Override
-  public void visitNode(AstNode astNode) {
-    if (!hasVisibility(astNode)) {
-      getContext().createLineViolation(this, "Explicitly mention the visibility of this {0}.", astNode, getMessageForMethodName(astNode));
+  // fixme (Lena) : could be refactored with https://github.com/pynicolas/sonar-php/commit/c8ba74d43c0816871e928d9415da68791fbde5e8
+  private void checkMethod(MethodDeclarationTree method, String className) {
+    if (!hasVisibilityModifier(method)) {
+      String methodName = method.name().text();
+      String message = String.format(MESSAGE, getMethodKind(methodName, className), methodName);
+      context().newIssue(KEY, message).tree(method);
     }
   }
 
-  private boolean hasVisibility(AstNode astNode) {
-    for (AstNode modifier : astNode.getChildren(PHPGrammar.MEMBER_MODIFIER)) {
-      if (VISIBILITIES.contains(modifier.getFirstChild().getType())) {
+  private boolean hasVisibilityModifier(MethodDeclarationTree method) {
+    for (SyntaxToken modifier : method.modifiers()) {
+      if (VISIBILITIES.contains(modifier.text().toLowerCase())) {
         return true;
       }
     }
     return false;
   }
 
-  private String getMessageForMethodName(AstNode methodNode) {
-    StringBuilder builder = new StringBuilder();
-    String name = methodNode.getFirstChild(PHPGrammar.IDENTIFIER).getTokenOriginalValue();
-
-    if (isConstructor(methodNode, name)) {
-      builder.append("constructor ");
-    } else if ("__destruct".equals(name)) {
-      builder.append("destructor ");
+  private String getMethodKind(String methodName, String className) {
+    if (methodName.equalsIgnoreCase("__construct") || methodName.equalsIgnoreCase(className)) {
+      return "constructor";
+    } else if (methodName.equalsIgnoreCase("__destruct")) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Move the "__destruct" string literal on the left side of this string comparison. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)